text/0000-ptr-meta.md

```diff
@@ -0,0 +1,387 @@
+* A `Thin` [trait alias].
+  If this RFC is implemented before type aliases are,
+  uses of `Thin` should be replaced with its definition.
+* A `metadata` free function
+* A `DynMetadata` struct
+* A `from_raw_parts` constructor for each of `*const T` and `*mut T`
+
+The bounds on `null()` and `null_mut()` function in that same module
+as well as the `NonNull::dangling` constructor
+are changed from (implicit) `T: Sized` to `T: ?Sized + Thin`.
+Similarly for the `U` type parameter of the `NonNull::cast` method.
+This enables using those functions with [extern types].
+
+The `Pointee` trait is implemented for all types.
+This can be relied on in generic code,
+even if a type parameter `T` does not have an explicit `T: Pointee` bound.
+This is similar to how the `Any` trait can be used without an explicit `T: Any` bound,
+only `T: 'static`, because a blanket `impl<T: 'static> Any for T {…}` exists.
+(Except that `Pointee` is not restricted to `'static`.)
+
+For the purpose of pointer casts being allowed by the `as` operator,
+a pointer to `T` is considered to be thin if `T: Thin` instead of `T: Sized`.
+This similarly includes extern types.
+
+`std::raw::TraitObject` and `std::raw` are deprecated and eventually removed.
+
+[trait alias]: https://github.com/rust-lang/rust/issues/41517
+[extern types]: https://github.com/rust-lang/rust/issues/43467
+
+```rust
+/// This trait is automatically implement for every type.
+///
+/// Raw pointer types and reference types in Rust can be thought of as made of two parts:
+/// a data pointer that contains the memory address of the value, and some metadata.
+///
+/// For statically-sized types (that implement the `Sized` traits)
+/// as well as for `extern` types,
+/// pointers are said to be “thin”: metadata is zero-sized and its type is `()`.
+///
+/// Pointers to [dynamically-sized types][dst] are said to be “fat”
+/// and have non-zero-sized metadata:
+///
+/// * For structs whose last field is a DST, metadata is the metadata for the last field
+/// * For the `str` type, metadata is the length in bytes as `usize`
+/// * For slice types like `[T]`, metadata is the length in items as `usize`
+/// * For trait objects like `dyn SomeTrait`, metadata is [`DynMetadata`].
+///
+/// In the future, the Rust language may gain new kinds of types
+/// that have different pointer metadata.
+///
+/// Pointer metadata can be extracted from a pointer or reference with the [`metadata`] function.
+/// The data pointer can be extracted by casting a (fat) pointer
+/// to a (thin) pointer to a `Sized` type the `as` operator,
+/// for example `(x: &dyn SomeTrait) as *const SomeTrait as *const ()`.
+///
+/// [dst]: https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts
+#[lang = "pointee"]
+pub trait Pointee {
+    /// The type for metadata in pointers and references to `Self`.
+    type Metadata: Copy + Send + Sync + Ord + Hash + Unpin + 'static;
```

Comment:
Having to list `Unpin` here is concerning... does that mean that if we get another auto trait in libstd we'll want to add it here but cannot due to backwards compatibility?

Reply:
Why are the trait bounds there in the first place?  I can understand `Copy` and maybe `Send + Sync`, since you copy the metadata when you copy the pointer.  But why does metadata need to be `Ord` or `Hash`?  If some function really needs those, it could always use `T: Pointee<Metadata: Ord>` or similar.

Reply:
I've explained the rationale of this list in https://github.com/rust-lang/rfcs/pull/2580#discussion_r228722830 and `Unpin` in https://github.com/rust-lang/rfcs/pull/2580#discussion_r232457570.

@comex Because the following works on stable all the way back to 1.0:

```rust
fn is_before<T: ?Sized>(a: *const T, b: *const T) -> bool {
    a < b
}
```

Without introducing `?DynSized`, the `T: ?Sized` have to match _all_ types including slices and trait pointers and `extern type` and custom DST pointers. The fat-pointer comparison is always implemented as `(a.ptr, a.meta) < (b.ptr, b.meta)` without requiring `T::Metadata: PartialOrd` in `is_before`. This means such bound must be implicitly satisfied everywhere, and thus included in the list.

Every other bound are included for the same reason that `*const T` and/or `&T` unconditionally implements that trait.

@RalfJung we could tweak the auto trait rule so that it needs to check `T::Metadata` first before impl for `*const T`, `*mut T`, `&T` and `&mut T`, assuming auto-trait is not stabilized _before_ `Pointee` is implemented.

Reply:
> does that mean that if we get another auto trait in libstd we'll want to add it here but cannot due to backwards compatibility?

Does it?

This RFC proposes that this trait is automatically implemented for all types, so "manual" implementations of it cannot exist. (Unless some day the language grows types that cannot be used behind a pointer, but I can’t imagine how that would be useful. And even then, I assume we’d disallow manual impls.)

And it’s only those impls that would potentially be broken by adding a new bound, right?

Reply:
> The fat-pointer comparison is always implemented as (a.ptr, a.meta) < (b.ptr, b.meta)

Is it really? I would expect it to be implemented as `a.ptr < b.ptr`, ignoring the metadata. That seems like a more reasonable implementation?

> This RFC proposes that this trait is automatically implemented for all types, so "manual" implementations of it cannot exist.

I see. That leaves no room for custom DST though. I am aware those are out-of-scope, but they should still be possible to add in a future-compatible way I assume.

Reply:
> I would expect it to be implemented as `a.ptr < b.ptr`

```rust
fn is_before<T: ?Sized>(a: *const T, b: *const T) -> bool {
    a < b
}
#[test]
fn test() {
    let a = &[1,2,3];
    assert!(is_before(&a[..1], &a[..2])); // because 1 < 2
    assert!(!is_before(&a[..2], &a[..1])); // because 2 > 1
}
```

Reply:
> That leaves no room for custom DST though.

Oh, that’s a good point. I suppose when custom DSTs are added, then we may need to freeze those bounds.

> The fat-pointer comparison is always implemented as `(a.ptr, a.meta) < (b.ptr, b.meta)`

Indeed, it looks like fat-pointer comparison is represented specifically in MIR, and lowered to LLVM integer-comparison instructions here:

https://github.com/rust-lang/rust/blob/1e2a73867/src/librustc_codegen_ssa/mir/rvalue.rs#L630-L642

It doesn’t call `PartialOrd` methods. In fact it’s the other way around: `PartialOrd` uses the `<` operator which is special-cased for raw pointers to not call the trait:

https://github.com/rust-lang/rust/blob/1e2a73867/src/libcore/ptr/mod.rs#L2930

So maybe custom DSTs would also need language-level changes in this area.

Reply:
`is_before` is a very good example for why it needs to be comparable etc. But I'm missing why it needs to be `Send` and `Sync` (I mean if the DST type is Send+Sync then the meta must be Send+Sync, too. But if not it also doesn't need to have that bounds I think).

Reply:
Today we have `impl<T: ?Sized> !Send for *mut T {}` (similarly `Sync`) in libcore, without a `T: Send` bound. For a wide pointer to be `Send`, the metadata needs to be `Send` as well.